topi/python/topi/cuda/nms.py

```diff
@@ -5,427 +5,66 @@
-                                             == is_descend)):
-                        p_data[tid] = data_new[tid-1]
-                        index_new[tid] = index_out[tid-1]
-                    with ib.else_scope():
-                        p_data[tid] = data_new[tid]
-                        index_new[tid] = index_out[tid]
-        with ib.if_scope(fshape % 2 == 1):
-            with ib.if_scope(tid < 1):
-                with ib.for_range(0, fshape, name="k") as k:
-                    index_out[tid] = index_new[tid]
-    body = ib.get()
-    return body
-
-
-def sort_ir_out(data, index, new_index, loc, output, axis_mul_before, axis_mul_after, axis):
-    """Low level IR routing subfunction 4/4 for writing sorted indices to output format.
-
-    Parameters
-    ----------
-    data: Buffer
-        Buffer of output boxes with class and score.
-
-    index : Buffer
-        Buffer of number of valid output boxes.
-
-    new_index : Buffer
-        Buffer of sorted indices in a flatten format.
-
-    loc : Buffer
-        Buffer of start locations of each sorting segment.
-
-    output : Buffer
-        Output buffer of output box indexes sorted by score.
-
-    axis_mul_before : int
-        The multiplication result of axis dimensions before axis.
-
-    axis_mul_after : int
-        The multiplication result of axis dimensions after axis.
-
-    axis : int
-        The axis used for sorting.
-
-    is_descend : bool
-        If the sorted data is in descending order.
-
-    Returns
-    -------
-    stmt : Stmt
-        The result IR statement.
-    """
-    max_threads = int(
-        tvm.target.current_target(allow_none=False).max_num_threads)
-    tx = tvm.thread_axis("threadIdx.x")
-    bx = tvm.thread_axis("blockIdx.x")
-    ib = tvm.ir_builder.create()
-    dshape = tvm.max(loc.shape[0], data.shape[axis])
-    p_index = ib.buffer_ptr(index)
-    index_new = ib.buffer_ptr(new_index)
-    sizes = ib.buffer_ptr(loc)
     p_out = ib.buffer_ptr(output)
     nthread_tx = max_threads
-    nthread_bx = dshape // max_threads + 1
+    nthread_bx = num_anchors // max_threads + 1
+    tx = tvm.thread_axis("threadIdx.x")
+    bx = tvm.thread_axis("vthread")
     ib.scope_attr(tx, "thread_extent", nthread_tx)
-    ib.scope_attr(bx, "thread_extent", nthread_bx)
-    tid = bx * max_threads + tx
-
-    with ib.if_scope(axis_mul_before * axis_mul_after > 1):
-        with ib.if_scope(tid < axis_mul_before * axis_mul_after):
-            i = tid / axis_mul_after
-            j = tid % axis_mul_after
-            base_idx = i * data.shape[axis] * axis_mul_after + j
-            with ib.for_range(0, data.shape[axis], name="k") as k:
-                with ib.if_scope(tid == 0):
-                    start = 0
-                with ib.else_scope():
-                    start = sizes[tid-1]
-                p_out[base_idx + k * axis_mul_after] = tvm.if_then_else(
-                    k < p_index[tid], index_new[k+start], k)
```

Comment:
@Laurawly still confused, if batch > 1, it should enter this if branch (since `axis_mul_before * axis_mul_after > 1`). Does `p_index[tid]` here mean that each batch has a different valid count?

Reply:
@vinx13 From here https://github.com/dmlc/tvm/blob/master/topi/python/topi/cuda/nms.py#L368 `axis` is always 1, so `axis_mul_before` and `axis_mul_after` are both 1.

Reply:
@Laurawly since `ndim == 2, axis == 1`, the actual loop is like
```
for i in range(0, 2):
  if i < 1:
     axis_mul_before *= data.shape[i]
```
I assume that `axis_mul_after == 1`, `axis_mul_before == data.shape[0]`, which is batch size, right?

Reply:
@vinx13 Yeah, that's right. I see what you mean. So each batch could have a different valid count when batch size > 1.